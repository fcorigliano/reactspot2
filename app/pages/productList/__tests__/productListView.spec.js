const React = require('react');
const ProductListView = require('../view');
const { render, screen, fireEvent, act } = require('@testing-library/react');
const restclient = require('nordic/restclient');
const mockProducts = require('./sample.json');

jest.mock('nordic/restclient', () => () => ({
    get: jest.fn(() => Promise.resolve({data: mockProducts}))
}));

describe('La view de ProductList', () => {
    let component;
    const i18n = { gettext: text => text};
    const mockConsole = jest.spyOn(console, "log");
    const spyGet = jest.spyOn(restclient(), 'get');

    beforeEach(async() => {
        await act(async() => {
            component = render(<ProductListView i18n={i18n}/>)
        });
    });

    it('1) Renderiza', async() => {
        await act(async() => {
            const { asFragment } = component;
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it('2) Renderiza un listado de productos desde el client con las propiedades title, price y thumbnail', async() => {
        const cellphone1Title = screen.getAllByText(/samsung galaxy/i)[0];
        const cellphone1Price = screen.getByText(/38999/i);
        const cellphone1Img = screen.getAllByRole('img')[0];
        expect(cellphone1Title).toBeInTheDocument();
        expect(cellphone1Price).toBeInTheDocument();
        expect(cellphone1Img.src).toBe("http://http2.mlstatic.com/D_737913-MLA49433851234_032022-I.jpg");
    });

    it('3) Cada producto cuenta también con la propiedad `description`', () => {
        const description = screen.getAllByText(/Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 4 cámaras/i)[0];
        expect(description).toBeInTheDocument();
    })

    it('4) Renderiza un input y un botón para agregar cada producto al carrito', () => {
        const inputs = screen.getAllByRole('spinbutton');
        const buttons = screen.getAllByRole('button');
        expect(inputs.length).toBeGreaterThanOrEqual(6);
        expect(buttons.length).toBeGreaterThanOrEqual(6);
    });

    it('5) Agrega el producto a un array de productos seleccionados cuando se ingresa una cantidad y se presiona el botón', async() => {
        const input = screen.getAllByRole('spinbutton')[0];
        const button = screen.getAllByRole('button')[0];
        console.log(spyGet.mock)
        fireEvent.change(input, { target: { value: 2}});
        await act(async() => {
            fireEvent.click(button);
            expect(mockConsole).toHaveBeenCalled();
        });
    });
});