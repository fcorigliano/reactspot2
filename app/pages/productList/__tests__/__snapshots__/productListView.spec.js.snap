// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`La view de ProductList 1) Renderiza 1`] = `
<DocumentFragment>
  <div
    class="demo"
  >
    <h1>
      productList
    </h1>
    <ol>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A03 128 Gb Negro 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_737913-MLA49433851234_032022-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A03 128 Gb Negro 4 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $38999
          </h3>
          <form
            id="MLA1131686227"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Doble cámara y más detalles Sus 2 cámaras traseras de 48 Mpx/2 Mpx te permitirán tomar imágenes con más detalles y lograr efectos únicos como el famoso modo retrato de poca profundidad de campo. Además, el dispositivo cuenta con cámara frontal de 5 Mpx para que puedas sacarte divertidas selfies o hacer videollamadas. Más para ver Con su pantalla PLS de 6.5, disfrutá de colores intensos y mayor nitidez en todos tus contenidos. Mayor rendimiento Su memoria RAM de 4 GB permite que tu smartphone funcione de manera fluida y sin demoras al realizar distintas tareas, jugar o navegar. Batería de duración superior ¡Desenchufate! Con la súper batería de 5000 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas. Gran capacidad de almacenamiento Con su memoria interna de 128 GB podrás almacenar archivos y aplicaciones de gran tamaño sin necesidad de subirlos a la nube y aprovechar tus momentos sin conexión para darles el máximo uso.
          </p>
        </div>
      </li>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt="Motorola Edge 20 Lite 128 Gb Gris 6 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_833619-MLA48233725917_112021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Motorola Edge 20 Lite 128 Gb Gris 6 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $69999
          </h3>
          <form
            id="MLA1113718634"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 3 cámaras principales de tu equipo. Poné a prueba tu creatividad y jugá con la iluminación, diferentes planos y efectos para obtener grandes resultados. Además, el dispositivo cuenta con cámara frontal de 32 Mpx para que puedas sacarte divertidas selfies o hacer videollamadas. Capacidad y eficiencia Con su potente procesador y memoria RAM de 6 GB tu equipo alcanzará un alto rendimiento con gran velocidad de transmisión de contenidos y ejecutará múltiples aplicaciones a la vez sin demoras. Desbloqueo facial y dactilar Máxima seguridad para que solo vos puedas acceder al equipo. Podrás elegir entre el sensor de huella dactilar para habilitar el teléfono en un toque, o el reconocimiento facial que permite un desbloqueo hasta un 30% más rápido. Batería de duración superior ¡Desenchufate! Con la súper batería de 5000 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas.
          </p>
        </div>
      </li>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy S20 Fe 128 Gb Cloud Navy 6 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_995741-MLA47861036467_102021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy S20 Fe 128 Gb Cloud Navy 6 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $112999
          </h3>
          <form
            id="MLA933415694"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Pensado para vos El Samsung S20 FE posee el novedoso sistema operativo Android 10 que incorpora respuestas inteligentes y acciones sugeridas para todas tus aplicaciones. Entre sus diversas funcionalidades encontrarás el tema oscuro, navegación por gestos y modo sin distracciones, para que completes tus tareas mientras disfrutás al máximo tu dispositivo. Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 3 cámaras principales de tu equipo. Poné a prueba tu creatividad y jugá con la iluminación, diferentes planos y efectos para obtener grandes resultados. Perfecta para los amantes del plano detalle. Su zoom óptico te ofrecerá la posibilidad de realizar acercamientos sin que tus capturas pierdan calidad. Experiencia visual increíble Mirá tus series y películas favoritas con la mejor definición a través de su pantalla Super AMOLED de 6.5. Disfrutá de colores brillantes y detalles precisos en todos tus contenidos. Desbloqueo veloz con tu huella dactilar Con el sensor de huella dactilar, el acceso es seguro y podrás desbloquearlo automáticamente con un toque. Batería de duración superior ¡Desenchufate! Con la súper batería de 4500 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas.
          </p>
        </div>
      </li>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt=" Moto G22 128 Gb Cosmic Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_854793-MLA50262132804_062022-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
             Moto G22 128 Gb Cosmic Black 4 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $53000
          </h3>
          <form
            id="MLA1145986600"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 4 cámaras principales de tu equipo. Poné a prueba tu creatividad y jugá con la iluminación, diferentes planos y efectos para obtener grandes resultados. Mayor rendimiento Su memoria RAM de 4 GB permite que tu smartphone funcione de manera fluida y sin demoras al realizar distintas tareas, jugar o navegar. Desbloqueo facial y dactilar Máxima seguridad para que solo vos puedas acceder al equipo. Podrás elegir entre el sensor de huella dactilar para habilitar el teléfono en un toque, o el reconocimiento facial que permite un desbloqueo hasta un 30% más rápido. Batería de duración superior ¡Desenchufate! Con la súper batería de 5000 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas.
          </p>
        </div>
      </li>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A22 128 Gb Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_774406-MLA47777566692_102021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A22 128 Gb Black 4 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $92899
          </h3>
          <form
            id="MLA1130306961"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 4 cámaras principales de tu equipo. Poné a prueba tu creatividad y jugá con la iluminación, diferentes planos y efectos para obtener grandes resultados. Experiencia visual increíble Mirá tus series y películas favoritas con la mejor definición a través de su pantalla Super AMOLED de 6.4. Disfrutá de colores brillantes y detalles precisos en todos tus contenidos. Mayor rendimiento Su memoria RAM de 4 GB permite que tu smartphone funcione de manera fluida y sin demoras al realizar distintas tareas, jugar o navegar. Desbloqueo facial y dactilar Máxima seguridad para que solo vos puedas acceder al equipo. Podrás elegir entre el sensor de huella dactilar para habilitar el teléfono en un toque, o el reconocimiento facial que permite un desbloqueo hasta un 30% más rápido. Batería de duración superior ¡Desenchufate! Con la súper batería de 5000 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas.
          </p>
        </div>
      </li>
      <li
        class="card"
      >
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A32 128 Gb Awesome Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_922219-MLA48428094384_122021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A32 128 Gb Awesome Black 4 Gb Ram 
          </h4>
          <h3
            class="price"
          >
            $63999
          </h3>
          <form
            id="MLA918474630"
          >
            <label>
              Cantidad: 
            </label>
            <input
              type="number"
            />
            <button
              type="submit"
            >
              Agregar al carrito
            </button>
          </form>
          <p>
            Fotografía profesional en tu bolsillo Descubrí infinitas posibilidades para tus fotos con las 4 cámaras principales de tu equipo. Poné a prueba tu creatividad y jugá con la iluminación, diferentes planos y efectos para obtener grandes resultados. Experiencia visual increíble Mirá tus series y películas favoritas con la mejor definición a través de su pantalla Super AMOLED de 6.4. Disfrutá de colores brillantes y detalles precisos en todos tus contenidos. Mayor rendimiento Su memoria RAM de 4 GB permite que tu smartphone funcione de manera fluida y sin demoras al realizar distintas tareas, jugar o navegar. Desbloqueo veloz con tu huella dactilar Con el sensor de huella dactilar, el acceso es seguro y podrás desbloquearlo automáticamente con un toque. Batería de duración superior ¡Desenchufate! Con la súper batería de 5000 mAh tendrás energía por mucho más tiempo para jugar, ver series o trabajar sin necesidad de realizar recargas.
          </p>
        </div>
      </li>
    </ol>
  </div>
</DocumentFragment>
`;
